catalog/content/core/content-core-videothumbnailplugin/src/main/java/ddf/content/plugin/video/VideoThumbnailPlugin.java

```diff
@@ -0,0 +1,466 @@
+
+        limitFFmpegProcessesSemaphore = new Semaphore(MAX_FFMPEG_PROCESSES, true);
+    }
+
+    /**
+     * Deletes the directory that holds the FFmpeg binary.
+     * <p>
+     * Called by Blueprint.
+     */
+    public void destroy() {
+        if (ffmpegPath != null) {
+            final File ffmpegDirectory =
+                    new File(FilenameUtils.getFullPathNoEndSeparator(ffmpegPath));
+            if (!FileUtils.deleteQuietly(ffmpegDirectory)) {
+                ffmpegDirectory.deleteOnExit();
+            }
+        }
+    }
+
+    private void copyFFmpegBinary(final InputStream inputStream) throws IOException {
+        final File ffmpegBinary = new File(ffmpegPath);
+
+        if (!ffmpegBinary.exists()) {
+            FileUtils.copyInputStreamToFile(inputStream, ffmpegBinary);
+            if (!ffmpegBinary.setExecutable(true)) {
+                LOGGER.warn(
+                        "Couldn't make FFmpeg binary at {} executable. It must be executable by its owner for the plugin to work.",
+                        ffmpegPath);
+            }
+        }
+    }
+
+    @Override
+    public CreateResponse process(final CreateResponse input) throws PluginExecutionException {
+        // TODO: How to handle application/octet-stream?
+        final MimeType createdMimeType = input.getCreatedContentItem()
+                .getMimeType();
+        final MediaType createdMediaType = MediaType.create(createdMimeType.getPrimaryType(),
+                createdMimeType.getSubType());
+        if (!createdMediaType.is(MediaType.ANY_VIDEO_TYPE)) {
+            return input;
+        }
+
+        final CreateRequest request = input.getRequest();
+        final Map<String, Serializable> properties = request.getProperties();
+        final ContentItem contentItem = input.getCreatedContentItem();
+
+        createThumbnail(contentItem, properties);
+
+        return new CreateResponseImpl(request,
+                contentItem,
+                input.getResponseProperties(),
+                properties);
+    }
+
+    @Override
+    public UpdateResponse process(final UpdateResponse input) throws PluginExecutionException {
+        // TODO: How to handle application/octet-stream?
+        final MimeType updatedMimeType = input.getUpdatedContentItem()
+                .getMimeType();
+        final MediaType updatedMediaType = MediaType.create(updatedMimeType.getPrimaryType(),
+                updatedMimeType.getSubType());
+        if (!updatedMediaType.is(MediaType.ANY_VIDEO_TYPE)) {
+            return input;
+        }
+
+        final UpdateRequest request = input.getRequest();
+        final Map<String, Serializable> properties = request.getProperties();
+        final ContentItem contentItem = input.getUpdatedContentItem();
+
+        createThumbnail(contentItem, properties);
+
+        return new UpdateResponseImpl(request,
+                contentItem,
+                input.getResponseProperties(),
+                properties);
+    }
+
```

Comment:
Yep, there's definitely some stuff in there that could be shared. In fact, most of it could be if there weren't the separate `getCreatedContentItem()` and `getUpdatedContentItem()` methods. Oddly, all of the interfaces in `ddf.content.operation` that extend `Response` provide a method to get the associated content item, but it's not provided as a method in the `Response` interface.